torch/csrc/jit/codegen/cuda/scheduler/vectorize_helper.cpp

```diff
@@ -17,171 +18,1092 @@ namespace fuser {
-    int num_merged_domains) {
-  const auto ndims = domain.size();
-  IterDomain* merged_id = nullptr;
-  bool is_merge_done = false;
-  for (const auto i : c10::irange(num_merged_domains)) {
-    auto id = domain.at(ndims - 1 - i);
-    // broadcast are ignored
-    if (id->isBroadcast()) {
-      continue;
+std::multiset<int64_t> getAllFactors(const std::multiset<int64_t>& in_vals) {
+  std::multiset<int64_t> factored_vals;
+  for (auto val : in_vals) {
+    auto factors = computeFactors(val);
+    factored_vals.insert(factors.begin(), factors.end());
+  }
+  return factored_vals;
+}
+
+std::pair<std::multiset<int64_t>, std::multiset<int64_t>> removeCommonFactors(
+    const std::multiset<int64_t>& set1,
+    const std::multiset<int64_t>& set2) {
+  // Will iterate over this list, removing entries from the following lists
+  auto simplified_set1 = set1;
+  auto simplified_set2 = set2;
+
+  simplified_set1.erase(1);
+  simplified_set2.erase(1);
+
+  for (auto numerator_factor : set1) {
+    auto denominator_factor_it = simplified_set2.find(numerator_factor);
+    if (denominator_factor_it != simplified_set2.end()) {
+      simplified_set2.erase(denominator_factor_it);
+      simplified_set1.erase(simplified_set1.find(numerator_factor));
     }
-    if (merged_id == nullptr) {
-      merged_id = id;
-    } else {
-      auto id_inner = merged_id;
-      auto id_outer = id;
-      merged_id = IterDomain::merge(id_outer, id_inner);
-      is_merge_done = true;
+  }
+
+  return {simplified_set1, simplified_set2};
+}
+
+// Given factors made up of a product of integers over a product of integers,
+// simplify the factors. This should really likely use a better factorization
+// algorithm, however going to start simple here.
+std::pair<std::vector<int64_t>, std::vector<int64_t>> removeCommonFactors(
+    const std::vector<int64_t>& vec1,
+    const std::vector<int64_t>& vec2) {
+  std::multiset<int64_t> set1(vec1.begin(), vec1.end());
+  std::multiset<int64_t> set2(vec2.begin(), vec2.end());
+
+  set1.erase(1);
+  set2.erase(1);
+
+  // Initial removal of common factors
+  std::tie(set1, set2) = removeCommonFactors(set1, set2);
+
+  // If denominator is empty no reason to continue trying to remove factors
+  if (set2.empty()) {
+    return {{set1.begin(), set1.end()}, {}};
+  }
+
+  // Break sets up into largest set of smallest factors
+  set1 = getAllFactors(set1);
+  set2 = getAllFactors(set2);
+
+  // Remove common factors again
+  std::tie(set1, set2) = removeCommonFactors(set1, set2);
+
+  return {{set1.begin(), set1.end()}, {set2.begin(), set2.end()}};
+}
+
+std::pair<std::multiset<Val*>, std::multiset<Val*>> removeSameVals(
+    const std::multiset<Val*>& set1,
+    const std::multiset<Val*>& set2) {
+  // Will iterate over this list, removing entries from the following lists
+  auto simplified_set1 = set1;
+  auto simplified_set2 = set2;
+
+  for (auto fact1 : set1) {
+    auto fact2_it = simplified_set2.find(fact1);
```

Comment:
Not sure what the context of this function would be, but do we need to do the comparison possibly with actual values rather than just comparing pointers?

Reply:
WDYM?

Reply:
Actually don't remember why I asked this question, but I think I was wondering about the definition of sameness, so for example we could also use ExpressionEvaluator to find "same" values, but I assume just using object addresses is fine for this use case.

Reply:
Object addresses are used in this one. There's the other simplifier which uses explicit const values. This approach is processing compile time constant values separately from the symbolic, so the assumption is if this is used, it's only possible to do by `Val*` not by Expression Evaluation.